Add support for look-around patterns using fancy-regex

duckontheweb opened the pull request:
Uses [`fancy-regex`](https://docs.rs/fancy-regex/0.5.0/fancy_regex/) in place of [`regex`](https://docs.rs/regex/) for matching against all patterns in JSON schemas. 

`regex` is still used when [converting a RegEx for ECMA 262](https://github.com/duckontheweb/jsonschema-rs/blob/b689ac5b32c29db0a36cf7ad3890e78f0cea96e3/jsonschema/src/keywords/pattern.rs#L78) to take advantage of the `Regex::replace_all` method, which is not included in `fancy-regex`.  Since `fancy-regex` depends on `regex` anyway, this does not affect the final size of the compiled binary (thanks to [this commit](https://github.com/niklasfasching/check_timed_logs_fast/commit/fa2ec6ef77d3befd03d8674e0ceae24846b620a6) for the idea). 

I also removed the test from #214, since that pattern is supported in this PR.

Stranger6667 commented on jsonschema/src/keywords/pattern.rs:
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
-            if !self.pattern.is_match(item) {
-                return false;
+            if let Ok(is_match) = self.pattern.is_match(item) {
+                if !is_match {
+                    return false;
+                }
```
I think that it could be 

```suggestion
                return is_match
```

duckontheweb replied:
Yes, thanks!

Stranger6667 commented on jsonschema/src/keywords/pattern.rs:
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
         if let Value::String(item) = instance {
-            if !self.pattern.is_match(item) {
-                return false;
+            if let Ok(is_match) = self.pattern.is_match(item) {
```
What about the `Err` variant? I believe it should lead to `false`.

duckontheweb replied:
Yeah, I had a hard time reasoning out what would be expected here. I started typing out a response here and then came to same conclusion that it should lead to `false`. I'll update that.

Stranger6667 commented on jsonschema/src/keywords/pattern.rs:
```diff
@@ -43,21 +41,25 @@ impl Validate for PatternValidator {
-                    instance,
-                    self.original.clone(),
-                ));
+            if let Ok(is_match) = self.pattern.is_match(item) {
```
If `Err` is returned, we should indicate this backtracking problem to the user with a new `ValidationError` variant.

duckontheweb replied:
Yes, good idea. I'll add this.

Stranger6667 reviewed (changes requested):
I like the direction and it looks good to me :)

I think that it is important to add a new variant to the `ValidationErrorKind` enum that will indicate the backtracking issue and use it in `PatternValidator::validate` and other cases when patterns are used (`additionalProperties`)

Stranger6667 commented:
Maybe one extra thing - it will be nice to have a test case for `pattern` with backtracking

duckontheweb commented:
@Stranger6667 How would you like me to handle the commits to address these comments. Should I squash them with the existing commit?

Stranger6667 commented:
In the end, I plan to squash them into one, so feel free to push more :)

duckontheweb commented:
I think I addressed all of the suggestions above via the new commits. I also added matching and non-matching test cases to `pattern.rs`.

I had a thought that it would be good to add an integration test for the combination of `patternProperties`, `properties`, and `additionalProperties` that I mentioned in [this comment](https://github.com/Stranger6667/jsonschema-rs/issues/183#issuecomment-831473463), but I wasn't sure how to fit it in with the official JSON Schema test suite.

Stranger6667 commented on jsonschema/src/keywords/pattern.rs:
```diff
@@ -43,35 +41,44 @@ impl Validate for PatternValidator {
 impl ToString for PatternValidator {
     fn to_string(&self) -> String {
-        format!("pattern: {}", self.pattern)
+        format!("pattern: {:?}", self.pattern)
```
Hm, is it required?

duckontheweb replied:
These were required because `fancy_regex::Regex` implements `Debug` but not `Display`.

Stranger6667 replied:
Aha, got it :) then it obsoletes some of my other comments

Stranger6667 commented on jsonschema/src/keywords/pattern.rs:
```diff
@@ -110,14 +117,16 @@ fn convert_regex(pattern: &str) -> Result<Regex, regex::Error> {
     // There will be no overflow, because the minimum value is 65 (char 'A')
     ((captures
-        .index(0)
+        .get(0)
+        .map(|m| m.as_str())
+        .expect("index 0 to return the whole match")
```
As the type didn't change, why are these changes needed?

duckontheweb replied:
Good catch. This got left over from my initial attempt to use `fancy_regex::Captures` here. Now that we're continuing to use `regex` this is no longer necessary, so I'll revert it.

Stranger6667 commented on jsonschema/src/keywords/pattern_properties.rs:
```diff
@@ -80,7 +80,9 @@ impl ToString for PatternPropertiesValidator {
                 .iter()
-                .map(|(key, validators)| { format!("{}: {}", key, format_validators(validators)) })
+                .map(|(key, validators)| {
+                    format!("{:?}: {}", key, format_validators(validators))
```
Is the format change needed?

Stranger6667 commented on jsonschema/src/keywords/pattern_properties.rs:
```diff
@@ -148,7 +150,7 @@ impl Validate for SingleValuePatternPropertiesValidator {
     fn to_string(&self) -> String {
         format!(
-            "patternProperties: {{{}: {}}}",
+            "patternProperties: {{{:?}: {}}}",
```
Is the format change needed?

duckontheweb replied:
Same as above, this is required because `fancy_regex::Regex` implements `Debug` but not `Display`.

Stranger6667 commented:
Looks great to me! :) Thank you for contributing. I appreciate it :)

I'll check it one more time tomorrow and probably will release these changes in the new release :)

duckontheweb commented:
I'm not sure what to make of that `clippy` failure. It doesn't show up when I run `cargo clippy --all-targets --all-features -- -D warnings` locally, but I'm on a Mac. I'll try running in an Ubuntu container as well to see if I can replicate it.

duckontheweb commented:
> Looks great to me! :) Thank you for contributing. I appreciate it :)
> 
> I'll check it one more time tomorrow and probably will release these changes in the new release :)

No problem, thanks for bearing with me on this. It's my first contribution to a Rust project and it's been really educational to see how this project is set up.

Stranger6667 commented:
> I'm not sure what to make of that clippy failure. It doesn't show up when I run cargo clippy --all-targets --all-features -- -D warnings locally, but I'm on a Mac. I'll try running in an Ubuntu container as well to see if I can replicate it.

Hm, maybe it is connected to the recent Rust toolchain update :shrug: I'll apply a fix after merge :)

```
pub(crate) const fn backtrack_limit(
```

Stranger6667 commented:
> No problem, thanks for bearing with me on this. It's my first contribution to a Rust project and it's been really educational to see how this project is set up.

I am glad to hear that! :)
